### PR TITLE
Add wave* and some time* functions

### PR DESCRIPTION
**Pull request by jminer**
Is this a good way to do unions (`MMTIME`)? It is similar to how bindgen translated the union.

### CODE REVIEW

File: src/mmsystem.rs [hunk 1]
```diff
@@ -1,8 +1,230 @@
...
+pub const CALLBACK_EVENT: ::DWORD = 0x00050000;
+//497 (Win 7 SDK)
+pub const WAVERR_BADFORMAT: MMRESULT = WAVERR_BASE + 0;
+pub const WAVERR_STILLPLAYING: MMRESULT = WAVERR_BASE + 1;
+pub const WAVERR_UNPREPARED: MMRESULT = WAVERR_BASE + 2;
+pub const WAVERR_SYNC: MMRESULT = WAVERR_BASE + 3;
+pub const WAVERR_LASTERROR: MMRESULT = WAVERR_BASE + 3;
+#[repr(C)] #[derive(Clone, Copy, Debug)]
+pub struct HWAVEIN__;
+pub type HWAVEIN = *mut HWAVEIN__;
```
Comment (retep998) [inline src/mmsystem.rs]: Use `DECLARE_HANDLE!(HWAVEIN, HWAVEIN__);`

**Comment by retep998** [comment]
I'm not sure how I want to deal with untagged unions. Ideally Rust would get something like https://github.com/rust-lang/rfcs/pull/724 which would makes things way easier. I'm not a fan of those methods returning a `*mut` instead of a `&mut`, plus if all you have is an immutable `&` then you can't access those union members at all.

File: src/mmsystem.rs (hunk 1 above)
Comment (jminer) [inline src/mmsystem.rs]: Fixed `HWAVEIN` and `HWAVEOUT`.

**Comment by jminer** [comment]
Yeah, I wish there were language support for C unions. What do you think about having `mut` and non-`mut` versions of each member?

``` rust
    pub unsafe fn ms<'a>(&'a self) -> &'a ::DWORD {
        ::std::mem::transmute(&self.data)
    }
    pub unsafe fn ms_mut<'a>(&'a mut self) -> &'a mut ::DWORD {
        ::std::mem::transmute(&mut self.data)
    }
```

The methods should probably have an `inline` attribute too.

**Comment by retep998** [comment]
There shouldn't be any need to explicitly specify the lifetime in this case as lifetime inference should be enough.

**Comment by jminer** [comment]
Right, thanks. I was thinking that `transmute` doesn't preserve the lifetime, but I forgot that's irrelevant as the body of a method doesn't affect lifetime elision. I updated the commit.

**Comment by retep998** [comment]
Do you think you could update this PR to try to make use of the `UNION!` macro I wrote?

**Comment by jminer** [comment]
Sure, it's updated. I think using a macro is a good improvement.

**Pull request merged**